File: 16 3Sum Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int n = nums.length;        
        int ans = 0;
        int diff = Integer.MAX_VALUE;
        Arrays.sort(nums);        
        for(int i=0;i<n-2;i++){        
            int l = i+1;
            int r = n-1;
            while(l<r){        
                int sum = nums[i] + nums[l] + nums[r];                
                if(Math.abs(sum-target)<diff) {                    
                    ans = sum;
                    diff = Math.abs(sum - target);
                }                
                if(sum==target){                    
                    return sum;                    
                } else if(sum > target){                    
                    r--;
                }
                else l++;
            }
        }
        return ans;
    }
}

// Runtime: 3 ms, faster than 98.88% of Java online submissions for 3Sum Closest.
// Memory Usage: 38.5 MB, less than 73.14% of Java online submissions for 3Sum Closest.

File: 16 3Sum Closest.c


int cmp(void* a,void* b){
    return *(int*)a - *(int*)b;
}
int abs(int a){
    return a>0?a:-a;
}

int threeSumClosest(int* nums, int numsSize, int target){
    qsort(nums,numsSize,sizeof(int),cmp);
    int ret = 0;
    int grap = INT_MAX;
    int** map = (int**)malloc(numsSize*sizeof(int*));
    for(int k=0;k<numsSize;k++){
        map[k] = (int*)calloc(numsSize,sizeof(int));
        for(int i=1; i+k<numsSize; i++){
            map[k][i+k] = nums[k] + nums[i+k];
        }
    }
    
    int i = 0;
    int j = 0;
    int k = 0;
    for(i=0; i<numsSize; i++){
        for(j=i+1; j<numsSize-1; j++){
            for(k=j+1; k<numsSize; k++){
                int temp = map[i][j] + nums[k];
                if(grap > abs(temp-target)) {
                    grap = abs(temp-target);
                    ret = temp;
                }
                if(temp > target){
                    break;
                }
            }
        }
    }
    return ret;
}


// 131 / 131 test cases passed.
// Status: Accepted
// Runtime: 140 ms
// Memory Usage: 20.1 MB
// Submitted: 0 minutes ago
